homeassistant/components/media_player/device_action.py

```diff
@@ -0,0 +1,163 @@
+    CONF_MEDIA_PLAY,
+    CONF_MEDIA_PLAY_PAUSE,
+    CONF_MEDIA_PREVIOUS_TRACK,
+    CONF_MEDIA_STOP,
+    CONF_SUPPORTED_FEATURES,
+    CONF_VOLUME_DOWN,
+    CONF_VOLUME_UP,
+    SERVICE_CLEAR_PLAYLIST,
+)
+
+ACTION_TYPES: Dict[str, Dict[str, Any]] = {
+    CONF_TURN_ON: {
+        CONF_SERVICE: SERVICE_TURN_ON,
+        CONF_SUPPORTED_FEATURES: SUPPORT_TURN_ON,
+    },
+    CONF_TURN_OFF: {
+        CONF_SERVICE: SERVICE_TURN_OFF,
+        CONF_SUPPORTED_FEATURES: SUPPORT_TURN_OFF,
+    },
+    CONF_TOGGLE: {
+        CONF_SERVICE: SERVICE_TOGGLE,
+        CONF_SUPPORTED_FEATURES: SUPPORT_TURN_ON | SUPPORT_TURN_OFF,
+    },
+    CONF_VOLUME_UP: {
+        CONF_SERVICE: SERVICE_VOLUME_UP,
+        CONF_SUPPORTED_FEATURES: SUPPORT_VOLUME_SET | SUPPORT_VOLUME_STEP,
+    },
+    CONF_VOLUME_DOWN: {
+        CONF_SERVICE: SERVICE_VOLUME_DOWN,
+        CONF_SUPPORTED_FEATURES: SUPPORT_VOLUME_SET | SUPPORT_VOLUME_STEP,
+    },
+    CONF_MEDIA_PLAY_PAUSE: {
+        CONF_SERVICE: SERVICE_MEDIA_PLAY_PAUSE,
+        CONF_SUPPORTED_FEATURES: SUPPORT_PLAY | SUPPORT_PAUSE,
+    },
+    CONF_MEDIA_PLAY: {
+        CONF_SERVICE: SERVICE_MEDIA_PLAY,
+        CONF_SUPPORTED_FEATURES: SUPPORT_PLAY,
+    },
+    CONF_MEDIA_PAUSE: {
+        CONF_SERVICE: SERVICE_MEDIA_PAUSE,
+        CONF_SUPPORTED_FEATURES: SUPPORT_PAUSE,
+    },
+    CONF_MEDIA_STOP: {
+        CONF_SERVICE: SERVICE_MEDIA_STOP,
+        CONF_SUPPORTED_FEATURES: SUPPORT_STOP,
+    },
+    CONF_MEDIA_NEXT_TRACK: {
+        CONF_SERVICE: SERVICE_MEDIA_NEXT_TRACK,
+        CONF_SUPPORTED_FEATURES: SUPPORT_NEXT_TRACK,
+    },
+    CONF_MEDIA_PREVIOUS_TRACK: {
+        CONF_SERVICE: SERVICE_MEDIA_PREVIOUS_TRACK,
+        CONF_SUPPORTED_FEATURES: SUPPORT_PREVIOUS_TRACK,
+    },
+    CONF_CLEAR_PLAYLIST: {
+        CONF_SERVICE: SERVICE_CLEAR_PLAYLIST,
+        CONF_SUPPORTED_FEATURES: SUPPORT_CLEAR_PLAYLIST,
+    },
+}
+
+ACTION_SCHEMA = cv.DEVICE_ACTION_BASE_SCHEMA.extend(
+    {
+        vol.Required(CONF_TYPE): vol.In(list(ACTION_TYPES.keys())),
+        vol.Required(CONF_ENTITY_ID): cv.entity_domain(DOMAIN),
+    }
+)
+
+
+async def async_get_actions(hass: HomeAssistant, device_id: str) -> List[dict]:
+    """List device actions for Media Player devices."""
+    registry = await entity_registry.async_get_registry(hass)
+    actions = []
+
+    # Get all the integrations entities for this device
+    for entry in entity_registry.async_entries_for_device(registry, device_id):
+        if entry.domain != DOMAIN:
+            continue
+
+        state = hass.states.get(entry.entity_id)
+
+        if state:
+            supported_features = state.attributes.get(CONF_SUPPORTED_FEATURES, 0)
+        else:
+            supported_features = entry.supported_features
+
+        # Add actions for each entity that belongs to this integration
+        for action in ACTION_TYPES:
+            action_config = ACTION_TYPES[action]
+            if supported_features & action_config[CONF_SUPPORTED_FEATURES]:
```

Comment:
Unclear what you want to check here. Given the the supported_features can be a bitwise combination like "SUPPORT_TURN_ON | SUPPORT_TURN_OFF" this is somewhat unclear. The check now will allow TURN_ON or TURN_OFF. It's enough with one of the feature. Is this what is wanted, or are both required? I suggest changing this to:
```python
if (supported_features & action_config[CONF_SUPPORTED_FEATURES]) == action_config[CONF_SUPPORTED_FEATURES]
```

And if the OR condition is also required, allow a list of features for the conf:
```python
if (any((x & supported_feature) == x for x in action_config[CONF_SUPPORTED_FEATURES]))
```

That way it becomes clearer what is expected.

Reply:
I tried to do it same way services are registered, see here: https://github.com/home-assistant/core/blob/dev/homeassistant/components/media_player/__init__.py#L218
After digging a bit more I found that core helper is broken in the same way: https://github.com/home-assistant/core/blob/dev/homeassistant/helpers/service.py#L434
```python
# Skip entities that don't have the required feature.
    if required_features is not None and not any(
        entity.supported_features & feature_set for feature_set in required_features
    ):
        continue
```
It accepts list of required features, currently there is no difference whether you pass
`[SUPPORT_A | SUPPORT_B]` or `[SUPPORT_A, SUPPORT_B]`.

Should I fix this too? Impact of that change could be huge.

I will adjust my code to match service registration, changing volumes check to list.